File: backend/ast.cpp
#include <caml/mlvalues.h>
#include <caml/alloc.h>
#include <caml/custom.h>
#include <caml/memory.h>
#include <caml/hash.h>
#include <ast.h>

// Assumes you've included your structs above
Sint convert_sint(value v) {
    switch (Tag_val(v)) {
        case 0: return Sint::Ti8;
        case 1: return Sint::Ti16;
        case 2: return Sint::Ti32;
        case 3: return Sint::Ti64;
        case 4: return Sint::Ti128;
        default: throw std::runtime_error("Unknown Signed Int variant");
    }
}

Uint convert_uint(value v) {
    switch (Tag_val(v)) {
        case 0: return Uint::Tu8;
        case 1: return Uint::Tu16;
        case 2: return Uint::Tu32;
        case 3: return Uint::Tu64;
        case 4: return Uint::Tu128;
        default: throw std::runtime_error("Unknown Unsigned Int variant");
    }
}

FloatTy convert_float_ty(value v) {
    switch (Tag_val(v)) {
        case 0: return FloatTy::Tf32;
        case 1: return FloatTy::Tf64;
        default: throw std::runtime_error("Unknown FloatTy variant");
    }
}

IntTy convert_int_ty(value v) {
    IntTy result;
    if (Tag_val(v) == 0) {
        result.tag = IntTyTag::Signed;
        result.sint = convert_sint(Field(v, 0));
    } else {
        result.tag = IntTyTag::Unsigned;
        result.uint = convert_uint(Field(v, 0));
    }
    return result;
}

RefTy convert_ref_ty(value v) {
    RefTy ref;

    switch (Tag_val(v)) {
        case 0: { // RString
            ref.tag = RefTyTag::RString;
            break;
        }
        case 1: { // RArray of ty
            ref.tag = RefTyTag::RArray;
            ref.inner = std::make_unique<Ty>(convert_ty(Field(v, 0)));
            break;
        }
        case 2: { // RFun of ty list * ret_ty
            ref.tag = RefTyTag::RFun;
            value tys = Field(v, 0);
            while (tys != Val_emptylist) {
                value head = Field(tys, 0);
                ref.args.push_back(convert_ty(head));
                tys = Field(tys, 1);
            }
            ref.ret = convert_ret_ty(Field(v, 1));
            break;
        }
    }

    return ref;
}

Ty convert_ty(value v) {
    Ty ty;

    switch (Tag_val(v)) {
        case 0: // TBool
            ty.tag = TyTag::TBool;
            break;

        case 1: // TInt of int_ty
            ty.tag = TyTag::TInt;
            ty.int_ty = std::make_unique<IntTy>(convert_int_ty(Field(v, 0)));
            break;

        case 2: // TFloat of float_ty
            ty.tag = TyTag::TFloat;
            ty.float_ty = std::make_unique<FloatTy>(convert_float_ty(Field(v, 0)));
            break;

        case 3: // TRef of ref_ty
            ty.tag = TyTag::TRef;
            ty.ref_ty = std::make_unique<RefTy>(convert_ref_ty(Field(v, 0)));
            break;
    }

    return ty;
}

RetTy convert_ret_ty(value v) {
    RetTy ret;

    switch (Tag_val(v)) {
        case 0: // RetVoid
            ret.tag = RetTyTag::RetVoid;
            break;

        case 1: // RetVal of ty
            ret.tag = RetTyTag::RetVal;
            ret.val = std::make_unique<Ty>(convert_ty(Field(v, 0)));
            break;
    }

    return ret;
}

UnOp convert_unop(value v) {
    switch (Tag_val(v)) {
        case 0: return UnOp::Neg;
        case 1: return UnOp::Not;
        default: throw std::runtime_error("Unknown UnOp");
    }
}

BinOp convert_binop(value v) {
    switch (Tag_val(v)) {
        case 0: return BinOp::Add;
        case 1: return BinOp::Sub;
        case 2: return BinOp::Mul;
        case 3: return BinOp::Div;
        case 4: return BinOp::At;
        case 5: return BinOp::Mod;
        case 6: return BinOp::Pow;
        case 7: return BinOp::Eqeq;
        case 8: return BinOp::Neq;
        case 9: return BinOp::Lt;
        case 10: return BinOp::Lte;
        case 11: return BinOp::Gt;
        case 12: return BinOp::Gte;
        case 13: return BinOp::And;
        case 14: return BinOp::Or;
        default: throw std::runtime_error("Unknown BinOp");
    }
}

Node<Exp> convert_exp_node(value v) {
    Node<Exp> node;
    // get node loc
    value exp = Field(v, 0);

    if (Is_block(exp)) {
        switch(Tag_val(exp)){
            case 0: {
                bool b = Bool_val(Field(exp, 0));
                node.elt.val = EBool{b};
                break;
            }
            case 1: {
                int i = Int_val(Field(exp, 0));
                node.elt.val = EInt{i};
                break;
            }
            case 2: {
                double d = Double_val(Field(exp, 0));
                node.elt.val = EFloat{d};
                break;
            }
            case 3: {
                std::string s = String_val(Field(exp, 0));
                node.elt.val = EStr{s};
                break;
            }
            case 4: {
                value id_node = Field(exp, 0);
                std::string var = String_val(Field(id_node, 0));
                node.elt.val = EVar{var};
                break;
            }
            case 5: {
                Node<Exp> callee = convert_exp_node(Field(exp, 0));
                std::vector<std::unique_ptr<Node<Exp>>> args;
                value arg_list = Field(exp, 1);
                while (arg_list != Val_emptylist) {
                    value arg = Field(arg_list, 0);
                    args.push_back(std::make_unique<Node<Exp>>(convert_exp_node(arg)));
                    arg_list = Field(arg_list, 1);
                }
                node.elt.val = ECall {
                    .callee = std::make_unique<Node<Exp>>(std::move(callee)),
                    .args = std::move(args),
                };
                break;
            }
            case 6: {
                BinOp bop = static_cast<BinOp>(convert_binop(Int_val(Field(exp, 0))));
                Node<Exp> lhs = convert_exp_node(Field(exp, 1));
                Node<Exp> rhs = convert_exp_node(Field(exp, 2));
                node.elt.val = EBop {
                    .op = bop,
                    .left = std::make_unique<Node<Exp>>(std::move(lhs)),
                    .right = std::make_unique<Node<Exp>>(std::move(rhs)),
                };
                break;
            }
            case 7: {
                UnOp uop = static_cast<UnOp>(convert_unop(Int_val(Field(exp, 0))));
                Node<Exp> operand = convert_exp_node(Field(exp, 1));
                node.elt.val = EUop {
                    .op = uop,
                    .arg = std::make_unique<Node<Exp>>(std::move(operand)),
                };
                break;
            }
            case 8: {
                Node<Exp> collection = convert_exp_node(Field(exp, 0));
                Node<Exp> idx = convert_exp_node(Field(exp, 1));
                node.elt.val = EIndex {
                    .collection = std::make_unique<Node<Exp>>(std::move(collection)),
                    .index = std::make_unique<Node<Exp>>(std::move(idx)),
                };
                break;
            }
            case 9: {
                std::vector<std::unique_ptr<Node<Exp>>> elements;
                value elements_list = Field(exp, 0);
                while (elements_list != Val_emptylist) {
                    value element = Field(elements_list, 0);
                    elements.push_back(std::make_unique<Node<Exp>>(convert_exp_node(element)));
                    elements_list = Field(elements_list, 1);
                }
                node.elt.val = EArray {
                    .elements = std::move(elements),
                };
                break;
            }
            case 10: {
                Node<Exp> start = convert_exp_node(Field(exp, 0));
                Node<Exp> end = convert_exp_node(Field(exp, 1));
                bool inclusive = Bool_val(Field(exp, 2));
                node.elt.val = ERange {
                    .start = std::make_unique<Node<Exp>>(std::move(start)),
                    .end = std::make_unique<Node<Exp>>(std::move(end)),
                    .inclusive = inclusive,
                };
                break;
            }
        }
    }
}

Node<Stmt> convert_stmt_node(value v) {
    Node<Stmt> node;
    // get node loc
    value stmt = Field(v, 0);

    if (Is_block(stmt)) {
        switch(Tag_val(stmt)){
            case 0: { 
                Node<Exp> lhs = convert_exp_node(Field(stmt, 0));
                AOp op = static_cast<AOp>(Int_val(Field(stmt, 1)));
                Node<Exp> rhs = convert_exp_node(Field(stmt, 2));
                node.elt.val = Assn {
                    .lhs = std::make_unique<Node<Exp>>(std::move(lhs)),
                    .op = op,
                    .rhs = std::make_unique<Node<Exp>>(std::move(rhs)),
                };
                break;
            }
            case 1: { 
                value decl = Field(stmt, 0);
                std::string id = String_val(Field(decl, 0));
                // at this point we know the types of all vdecls (after typechecking)
                Ty ty = convert_ty(Field(decl, 1));
                Node<Exp> init = convert_exp_node(Field(decl, 2));
                bool is_const = Bool_val(Field(decl, 3));
                node.elt.val = VDecl {
                    .id = id,
                    .ty = std::move(ty),
                    .init = std::make_unique<Node<Exp>>(std::move(init)),
                    .is_const = is_const,
                };
                break;
            }
            case 2: { 
                value ret_val = Field(stmt, 0);
                if (Is_block(ret_val)) {
                    node.elt.val = Ret {
                        .value = std::make_unique<Node<Exp>>(convert_exp_node(Field(ret_val, 0))),
                    };
                } else {
                    node.elt.val = Ret {
                        .value = std::nullopt,
                    };
                }
                break;
            }
            case 3: { 
                Node<Exp> callee = convert_exp_node(Field(stmt, 0));
                std::vector<std::unique_ptr<Node<Exp>>> args;
                value arg_list = Field(stmt, 1);
                while (arg_list != Val_emptylist) {
                    value arg = Field(arg_list, 0);
                    args.push_back(std::make_unique<Node<Exp>>(convert_exp_node(arg)));
                    arg_list = Field(arg_list, 1);
                }
                node.elt.val = SCall {
                    .callee = std::make_unique<Node<Exp>>(std::move(callee)),
                    .args = std::move(args),
                };
                break;
            }
            case 4: { 
                Node<Exp> cond = convert_exp_node(Field(stmt, 0));
                std::vector<std::unique_ptr<Node<Stmt>>> then_branch;
                value then_list = Field(stmt, 1);
                while (then_list != Val_emptylist) {
                    value then_stmt = Field(then_list, 0);
                    then_branch.push_back(std::make_unique<Node<Stmt>>(convert_stmt_node(then_stmt)));
                    then_list = Field(then_list, 1);
                }
                std::vector<std::unique_ptr<Node<Stmt>>> else_branch;
                value else_list = Field(stmt, 2);
                while (else_list != Val_emptylist) {
                    value else_stmt = Field(else_list, 0);
                    else_branch.push_back(std::make_unique<Node<Stmt>>(convert_stmt_node(else_stmt)));
                    else_list = Field(else_list, 1);
                }
                node.elt.val = If {
                    .cond = std::make_unique<Node<Exp>>(std::move(cond)),
                    .then_branch = std::move(then_branch),
                    .else_branch = std::move(else_branch),
                };
                break;
            }
            case 5: { 
                Node<Exp> cond = convert_exp_node(Field(stmt, 0));
                std::vector<std::unique_ptr<Node<Stmt>>> body;
                value body_list = Field(stmt, 1);
                while (body_list != Val_emptylist) {
                    value body_stmt = Field(body_list, 0);
                    body.push_back(std::make_unique<Node<Stmt>>(convert_stmt_node(body_stmt)));
                    body_list = Field(body_list, 1);
                }
                node.elt.val = While {
                    .cond = std::make_unique<Node<Exp>>(std::move(cond)),
                    .body = std::move(body),
                };
                break;
            }
        }
    } else {
        switch(Tag_val(stmt)){
            case 6: { 
                node.elt.val = Break{};
                break;
            }
            case 7: { 
                node.elt.val = Continue{};
                break;
            }
        }
    }
}

Node<GDecl> convert_gdecl_node(value v) {
    Node<GDecl> node;
    // get node loc
    value gdecl_val = Field(v, 0);

    GDecl g;
    g.name = std::string(String_val(Field(gdecl_val, 0)));
    g.init = convert_exp_node(Field(gdecl_val, 1));

    node.elt = std::move(g);
    return node;
}

Node<FDecl> convert_fdecl_node(value v) {
    Node<FDecl> node;
    // get node loc
    value fdecl_val = Field(v, 0);

    FDecl f;
    f.rtyp = convert_ret_ty(Field(fdecl_val, 0));
    f.fname = std::string(String_val(Field(fdecl_val, 1)));

    value args = Field(fdecl_val, 2);
    while (args != Val_emptylist) {
        value pair = Field(args, 0);
        f.args.emplace_back(
            convert_ty(Field(pair, 0)),
            std::string(String_val(Field(pair, 1)))
        );
        args = Field(args, 1);
    }

    value body = Field(fdecl_val, 3);
    while (body != Val_emptylist) {
        f.body.push_back(convert_stmt_node(Field(body, 0)));
        body = Field(body, 1);
    }

    node.elt = std::move(f);
    return node;
}

Decl convert_decl(value v) {
    Decl decl;
    int tag = Tag_val(v);

    switch (tag) {
        case 0: 
            decl.val = convert_gdecl_node(Field(v, 0));
            break;
        case 1:
            decl.val = convert_fdecl_node(Field(v, 0));
            break;
        default:
    }

    return decl;
}

Program convert_program(value v) {
    Program program;
    value decls = Field(v, 0);
    while (decls != Val_emptylist) {
        value decl = Field(decls, 0);
        program.push_back(convert_decl(decl));
        decls = Field(decls, 1);
    }
    return program;
}

File: frontend/src/ast/ast.ml
module Range = Util.Range

type 'a node = { elt : 'a; loc : Range.t }

let show_node show_elt { elt; loc } =
  Printf.sprintf "{ elt = %s; loc = %s }" (show_elt elt)
    (Range.string_of_range loc)

let no_loc x = { elt = x; loc = Range.norange }

type id = string
type sint = Ti8 | Ti16 | Ti32 | Ti64 | Ti128 [@@deriving show]
type uint = Tu8 | Tu16 | Tu32 | Tu64 | Tu128 [@@deriving show]
type float_ty = Tf32 | Tf64 [@@deriving show]
type int_ty = TSigned of sint | TUnsigned of uint [@@deriving show]

type ty = TBool | TInt of int_ty | TFloat of float_ty | TRef of ref_ty
(* | TOpt of ty *)

and ref_ty =
  | RString
  | RArray of ty * Z.t
  | RClass of id
  | RFun of ty list * ret_ty
  | RGeneric of string * ty list

and ret_ty = RetVoid | RetVal of ty

type unop = Neg | Not [@@deriving show]

type binop =
  | Add
  | Sub
  | Mul
  | Div
  | At
  | Mod
  | Pow
  | Shl
  | Lshr
  | Ashr
  | Eqeq
  | Neq
  | Lt
  | Lte
  | Gt
  | Gte
  | And
  | Or
  | Xor (* | BXor 
  | BAnd 
  | BOr *)
[@@deriving show]

type aop =
  | Eq
  | PluEq
  | MinEq
  | TimEq
  | DivEq
  | AtEq
  | PowEq
  | ModEq
  | ShlEq
  | LShrEq
  | AShrEq
[@@deriving show]

type exp =
  | Bool of bool
  | Int of Z.t
  | Float of float
  | Str of string
  | Id of id
  | Proj of exp node * id
  | Call of exp node * exp node list
  | Bop of binop * exp node * exp node
  | Uop of unop * exp node
  | Index of exp node * exp node
  | Array of exp node list
  | Cast of exp node * ty
  | ObjInit of id node * (id node * exp node) list
  | Lambda of id list * block
  | TypedLambda of (id * ty) list * ret_ty * block
  | Null

and vdecl = id * ty option * exp node option * bool
and ldecl = id node * ref_ty option * exp node

and stmt =
  | Assn of exp node * aop * exp node
  | LambdaDecl of ldecl
  | Decl of vdecl (* includes whether it was declared as constant or not *)
  | Ret of exp node option
  | SCall of exp node * exp node list
  | If of exp node * block * block
  | ForEach of id node * exp node * block (* no step in for-each *)
  | For of
      (* iterator, start, stop, incl, step, body *)
      id node
      * (exp node * exp node * bool)
      * exp node option
      * block
  | While of exp node * block
  | Break
  | Continue

and block = stmt node list

type gdecl = { name : id; init : exp node }

type fdecl = {
  frtyp : ret_ty;
  fname : id;
  args : (ty * id) list;
  mutable body : block;
}

type field = { fieldName : id; ftyp : ty; init : exp node option }

type cdecl = {
  cname : id;
  impls : id list;
  fields : vdecl node list;
  methods : fdecl node list;
}

type program = Prog of fdecl node list * cdecl node list

(* Utility for indentation *)
let indent n = String.make (n * 2) ' '

(* Base show functions *)

let show_sint = function
  | Ti8 -> "i8"
  | Ti16 -> "i16"
  | Ti32 -> "i32"
  | Ti64 -> "i64"
  | Ti128 -> "i128"

let show_uint = function
  | Tu8 -> "u8"
  | Tu16 -> "u16"
  | Tu32 -> "u32"
  | Tu64 -> "u64"
  | Tu128 -> "u128"

let show_float_ty = function Tf32 -> "Tf32" | Tf64 -> "Tf64"

let show_int_ty = function
  | TSigned s -> Printf.sprintf "TSigned(%s)" (show_sint s)
  | TUnsigned u -> Printf.sprintf "TUnsigned(%s)" (show_uint u)

(* Type printers *)

let rec show_ref_ty ?(lvl = 0) = function
  | RString -> "RString"
  | RArray (t, sz) ->
      Printf.sprintf "RArray(%s, %s)"
        (show_ty ~lvl:(lvl + 1) t)
        (Z.to_string sz)
  | RClass cn -> Printf.sprintf "RClass(%s)" cn
  | RFun (args, ret) ->
      let args_s =
        String.concat "; " (List.map (fun a -> show_ty ~lvl:(lvl + 1) a) args)
      in
      Printf.sprintf "RFun([\n%s%s\n%s], %s)"
        (indent (lvl + 1))
        args_s (indent lvl)
        (show_ret_ty ~lvl:(lvl + 1) ret)
  | RGeneric _ -> failwith "generics not supported yet"

and show_ret_ty ?(lvl = 0) = function
  | RetVoid -> "RetVoid"
  | RetVal t -> Printf.sprintf "RetVal(%s)" (show_ty ~lvl:(lvl + 1) t)

and show_ty ?(lvl = 0) = function
  | TBool -> "TBool"
  | TInt it -> Printf.sprintf "TInt(%s)" (show_int_ty it)
  | TFloat ft -> Printf.sprintf "TFloat(%s)" (show_float_ty ft)
  | TRef rt -> Printf.sprintf "TRef(%s)" (show_ref_ty ~lvl:(lvl + 1) rt)
(* | TOpt t -> Printf.sprintf "TOpt(%s)" (show_ty t) *)

let rec show_exp ?(lvl = 0) = function
  | Bool b -> Printf.sprintf "Bool(%b)" b
  | Int i -> Printf.sprintf "Int(%s)" (Z.to_string i)
  | Float f -> Printf.sprintf "Float(%f)" f
  | Null -> Printf.sprintf "Null"
  | Str s -> Printf.sprintf "Str(%S)" s
  | Id id -> Printf.sprintf "Id(%s)" id
  | Call (fn, args) ->
      let args_s =
        String.concat ";\n"
          (List.map
             (fun a ->
               Printf.sprintf "%s%s" (indent (lvl + 2)) (show_node show_exp a))
             args)
      in
      Printf.sprintf "%sCall(\n%s%s,\n%s[\n%s\n%s])" (indent lvl)
        (indent (lvl + 1))
        (show_node show_exp fn)
        (indent (lvl + 1))
        args_s (indent lvl)
  | Bop (op, lhs, rhs) ->
      Printf.sprintf "%sBop(%s,\n%s%s,\n%s%s)" (indent lvl) (show_binop op)
        (indent (lvl + 1))
        (show_node show_exp lhs)
        (indent (lvl + 1))
        (show_node show_exp rhs)
  | Uop (op, e) ->
      Printf.sprintf "%sUop(%s, %s)" (indent lvl) (show_unop op)
        (show_node show_exp e)
  | Index (arr, idx) ->
      Printf.sprintf "%sIndex(%s, %s)" (indent lvl) (show_node show_exp arr)
        (show_node show_exp idx)
  | Array elems ->
      let elems_s =
        String.concat ";\n"
          (List.map
             (fun e ->
               Printf.sprintf "%s%s" (indent (lvl + 2)) (show_node show_exp e))
             elems)
      in
      Printf.sprintf "%sArray([\n%s\n%s])" (indent lvl) elems_s (indent lvl)
  | Cast (e, t) ->
      Printf.sprintf "%sCast(%s, %s)" (indent lvl) (show_node show_exp e)
        (show_ty ~lvl:(lvl + 1) t)
  | Proj (e, i) ->
      Printf.sprintf "%sProj(%s, %s)" (indent lvl) (show_node show_exp e) i
  | ObjInit (cn, fields) ->
      let fields_s =
        String.concat ";\n"
          (List.map
             (fun (f, e) ->
               Printf.sprintf "%s%s = %s"
                 (indent (lvl + 2))
                 (show_node (fun x -> x) f)
                 (show_node show_exp e))
             fields)
      in
      Printf.sprintf "%sObjInit(%s, [\n%s\n%s])" (indent lvl)
        (show_node (fun x -> x) cn)
        fields_s (indent lvl)
  | Lambda (params, body) ->
      let params_s = String.concat ", " params in
      let body_s =
        String.concat ";\n" (List.map (show_node_stmt ~lvl:(lvl + 2)) body)
      in
      Printf.sprintf "%sLambda(params=[%s], body=[\n%s\n%s])" (indent lvl)
        params_s body_s (indent lvl)
  | TypedLambda (params, ret_ty, body) ->
      let params_s =
        String.concat "; "
          (List.map
             (fun (id, ty) -> Printf.sprintf "(%s : %s)" id (show_ty ty))
             params)
      in
      let body_s =
        String.concat ";\n" (List.map (show_node_stmt ~lvl:(lvl + 2)) body)
      in
      Printf.sprintf "%sTypedLambda(params=[%s]; ret=%s; body=[\n%s\n%s])"
        (indent lvl) params_s
        (show_ret_ty ~lvl:(lvl + 1) ret_ty)
        body_s (indent lvl)

(* Variable declarations *)

and show_vdecl ?(lvl = 0) (id, ty_opt, exp_opt, is_const) =
  Printf.sprintf "%sDecl{id=%s; ty=%s; const=%b;\n%sinit=%s}" (indent lvl) id
    (match ty_opt with Some ty -> show_ty ~lvl:(lvl + 1) ty | None -> "None")
    is_const
    (indent (lvl + 1))
    (if Option.is_some exp_opt then show_node show_exp (Option.get exp_opt)
     else "<default>")

and show_ldecl ?(lvl = 0) (id, ty_opt, exp) =
  Printf.sprintf "%sLambda{id=%s; ty=%s ;\n%sinit=%s}" (indent lvl)
    (show_node (fun x -> x) id)
    (match ty_opt with
    | Some ty -> show_ref_ty ~lvl:(lvl + 1) ty
    | None -> "None")
    (indent (lvl + 1))
    (show_node show_exp exp)
(* Statements *)

and show_stmt ?(lvl = 0) = function
  | Assn (lhs, op, rhs) ->
      Printf.sprintf "%sAssn(\n%s%s,\n%s%s,\n%s%s)" (indent lvl)
        (indent (lvl + 1))
        (show_node show_exp lhs)
        (indent (lvl + 1))
        (show_aop op)
        (indent (lvl + 1))
        (show_node show_exp rhs)
  | Decl v -> show_vdecl ~lvl v
  | LambdaDecl l -> show_ldecl ~lvl l
  | Ret exp_opt ->
      let e_s =
        match exp_opt with Some e -> show_node show_exp e | None -> "None"
      in
      Printf.sprintf "%sRet(%s)" (indent lvl) e_s
  | SCall (fn, args) ->
      let args_s =
        String.concat ";\n"
          (List.map
             (fun a ->
               Printf.sprintf "%s%s" (indent (lvl + 2)) (show_node show_exp a))
             args)
      in
      Printf.sprintf "%sSCall(%s, [\n%s\n%s])" (indent lvl)
        (show_node show_exp fn) args_s (indent lvl)
  | If (cond, then_block, else_block) ->
      Printf.sprintf "%sIf(\n%scond=%s,\n%sthen=[\n%s\n%s],\n%selse=[\n%s\n%s])"
        (indent lvl)
        (indent (lvl + 1))
        (show_node show_exp cond)
        (indent (lvl + 1))
        (String.concat ";\n"
           (List.map (show_node_stmt ~lvl:(lvl + 2)) then_block))
        (indent (lvl + 1))
        (indent (lvl + 1))
        (String.concat ";\n"
           (List.map (show_node_stmt ~lvl:(lvl + 2)) else_block))
        (indent lvl)
  | ForEach (id, start, body) ->
      Printf.sprintf "%sForEach(%s in %s) [\n%s\n%s]" (indent lvl)
        (show_node (fun x -> x) id)
        (show_node show_exp start)
        (String.concat ";\n" (List.map (show_node_stmt ~lvl:(lvl + 2)) body))
        (indent lvl)
  | For (id, (start, fin, incl), step_opt, body) ->
      Printf.sprintf "%sFor(%s from %s to %s incl=%b step=%s) [\n%s\n%s]"
        (indent lvl)
        (show_node (fun x -> x) id)
        (show_node show_exp start) (show_node show_exp fin) incl
        (match step_opt with Some s -> show_node show_exp s | None -> "None")
        (String.concat ";\n" (List.map (show_node_stmt ~lvl:(lvl + 2)) body))
        (indent lvl)
  | While (cond, body) ->
      Printf.sprintf "%sWhile(%s) [\n%s\n%s]" (indent lvl)
        (show_node show_exp cond)
        (String.concat ";\n" (List.map (show_node_stmt ~lvl:(lvl + 2)) body))
        (indent lvl)
  | Break -> Printf.sprintf "%sBreak" (indent lvl)
  | Continue -> Printf.sprintf "%sContinue" (indent lvl)

and show_node_stmt ?(lvl = 0) stmt_node =
  Printf.sprintf "%s;\n" (show_node (show_stmt ~lvl) stmt_node)

(* Function declarations *)

let show_fdecl ?(lvl = 0) { frtyp; fname; args; body } =
  let args_s =
    String.concat "; "
      (List.map
         (fun (ty, id) -> Printf.sprintf "(%s, %s)" (show_ty ty) id)
         args)
  in
  Printf.sprintf "%sfdecl{name=%s; ret=%s; args=[%s]; body=[\n%s\n%s]}"
    (indent lvl) fname
    (show_ret_ty ~lvl:(lvl + 1) frtyp)
    args_s
    (String.concat "" (List.map (show_node_stmt ~lvl:(lvl + 1)) body))
    (indent lvl)

(* Class and program printers *)

let show_field ?(lvl = 0) { fieldName; ftyp; init } =
  Printf.sprintf "%s%s: %s = %s" (indent lvl) fieldName
    (show_ty ~lvl:(lvl + 1) ftyp)
    (match init with Some e -> show_node show_exp e | None -> "None")

let show_cdecl ?(lvl = 0) { elt = { cname; impls; fields; methods }; loc = _ } =
  let fields_s =
    String.concat ";\n"
      (List.map (show_node (show_vdecl ~lvl:(lvl + 2))) fields)
  in
  let methods_s =
    String.concat "\n"
      (List.map (fun m -> show_fdecl ~lvl:(lvl + 2) m.elt) methods)
  in
  Printf.sprintf
    "%scdecl{name=%s; impls=[%s];\n%sfields=[\n%s\n%s];\n%smethods=[\n%s\n%s]}"
    (indent lvl) cname (String.concat ", " impls)
    (indent (lvl + 1))
    fields_s
    (indent (lvl + 1))
    (indent (lvl + 1))
    methods_s (indent lvl)

let show_decl ?(lvl = 0) d = show_fdecl ~lvl d.elt

let show_prog (Prog (fns, cns)) =
  let cns_s = String.concat "\n" (List.map (show_cdecl ~lvl:1) cns) in
  let fns_s = String.concat "\n" (List.map (show_decl ~lvl:1) fns) in
  Printf.sprintf "Program{\n%s\n%s\n}" cns_s fns_s
